compiler/src/main/scala/org/scalajs/nscplugin/GenJSCode.scala

```diff
@@ -6442,8 +6448,21 @@ abstract class GenJSCode[G <: Global with Singleton](val global: G)
 
         (f, true)
       } else {
-        val f = js.Select(qual, encodeClassName(sym.owner),
-            encodeFieldSym(sym))(toIRType(sym.tpe))
+        val className = encodeClassName(sym.owner)
+        val fieldIdent = encodeFieldSym(sym)
+
+        // #4370 Fields cannot have type NothingType
+        val f = toIRType(sym.tpe) match {
+          case jstpe.NothingType =>
+            val select = js.Select(qual, className, fieldIdent)(
+                encodeClassType(RuntimeNothingClass))
+            if (forAssign)
+              select
+            else
+              js.Throw(select) // this is similar to what Scala/JVM does in some places
```

Comment:
The more I think about this, the less it makes sense IMO (here and in the deser hack). Why would we throw the value of the field?

Instead of this, could we model this as boxing / unboxing (just like with `null`)? So the code here would be something like:

```scala
toIRType(sym.tpe) match {
  case jstpe.NothingType =>
    val f = js.Select(qual, className, fieldIdent)(
        encodeClassType(RuntimeNothingClass))
    (f, true) // unboxing `Nothing` will throw.

  case ftpe =>
    val f = js.Select(qual, className, fieldIdent)(ftpe)
    (f, false)
}
```

This would make the `forAssign` parameter unnecessary.

Doing the same thing in the deser hack is a bit nasty, because we'd have to generate a tree to throw a class cast exception, so I'm less sure about that.

Reply:
Yes, for `GenJSCode` that's a good idea, indeed. I've changed it as suggested.

For the deser hack, I still think the current write is better because it is Scala-agnostic, and easy to identify in `Assign`.